### PR TITLE
Test for proper serialization/deserialization of Hedera protobuf objects

### PR DESCRIPTION
**Pull request by deyanzz**
**Description**:
Introducing new tests to verify serialization & deserialization of Hedera protobuf objects.
Testing only methods `toProtobuf` and `fromProtobuf` in the following order:

---

**Related issue(s)**: [Issue #163](https://github.com/hashgraph/hedera-sdk-cpp/issues/163)

---

Fixes #
Added the following implementation files for each test suit

1. `ED25519PublicKeyTest.cc` - add test `ED25519PublicKeyToProtobuf`;
2. `ECDSAPublicKeyTest.cc` - add test `PublicKeyFromProtobuf`;
3. `ExchangeRateTest.cc` - add tests `DeserializeExchangeRateFromProtobuf` & 
`DeserializeExchangeRateSetFromProtobuf`;
4. `AccountBalanceTest.cc` - add test `DeserializeAccountBalanceFromProtobuf`;
5. `AccountIdTest.cc` - add tests `SerializeAccountIdToProtobuf` & `DeserializeAccountIdFromProtobuf`;
6. `ContractIdTest.cc` - add test `SerializeContractIdТoProtobuf` & `DeserializeContractIdFromProtobuf`;
7. `TransactionIdTest.cc` - add test `SerializeTransactionIdToProtobuf` & `DeserializeTransactionIdFromProtobuf`;
8. `TransferTest.cc` - add test `SerializeTransferToProtobuf` & `DeserializeTransferFromProtobuf`

Following classes already have tests for protobuf serialization / deserialization:
 - `TransactionReceipt`
 - `TransactionRecord`
 - `TransactionResponse`

The following classes couldn't be tested, because all of them seem to be internal:
 - class `Endpoint`
 - class `NodeAddress`
 - class `NodeAddressBook`
 - class `DurationConverter`
 - class `TimestampConverter`
---

**Notes for reviewer**:
These serialization / deserialization tests could be moved in the previously added test files to avoid adding new test suites named like `SerializationTests`, `DeserializationTests` and `SerializationAndDeserializationTests`.
For example:
 - tests `SerializeAccountIdToProtobuf` & `DeserializeAccountIdFromProtobuf` could be moved into `AccountId` as they are testing only class `AccountId`.
 - test `DeserializeTransactionRecordFromProtobuf` could be moved into `TransactionRecord` as they are testing only `TransactionRecord`.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  // When
+  auto protoAccountId = std::unique_ptr<proto::AccountID>(testAccountId.toProtobuf());
+
+  // Then
+  EXPECT_EQ(protoAccountId->shardnum(), testAccountId.getShardNum());
+  EXPECT_EQ(protoAccountId->realmnum(), testAccountId.getRealmNum());
+  EXPECT_EQ(protoAccountId->accountnum(), testAccountId.getAccountNum());
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeAccountIdFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:69]: To write this test I used test [ProtobufAccountId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/AccountIdTest.cc#L125) from `AccountIdTest`.

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  testProtoAccountId.set_accountnum(789);
+
+  // When
+  AccountId accountId = AccountId::fromProtobuf(testProtoAccountId);
+  
+  // Then
+  EXPECT_EQ(accountId.toString(), "123.456.789");
+}
+
+TEST_F(SerializationAndDeserializationTests, SerializeContractIdТoProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:84]: To write this test I used test [ProtobufContractId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ContractIdTest.cc#L76) from ContractIdTest.

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  // When
+  auto protoContractId = std::unique_ptr<proto::ContractID>(testContractId.toProtobuf());
+
+  // Then
+  EXPECT_EQ(protoContractId->shardnum(), testShardNum);
+  EXPECT_EQ(protoContractId->realmnum(), testRealmNum);
+  EXPECT_EQ(protoContractId->contractnum(), testContractNum);
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeContractIdFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:101]: To write this test I used test [ProtobufContractId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ContractIdTest.cc#L76) from ContractIdTest.

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+private:
+  const uint64_t mShardNum = 1;
+  const uint64_t mRealmNum = 2;
+  const uint64_t mContractNum = 3;
+  const AccountId mAccountId = AccountId(10ULL);
+  const std::chrono::system_clock::time_point mValidStart = std::chrono::system_clock::time_point();
+};
+
+TEST_F(SerializationAndDeserializationTests, SerializeAccountIdToProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:54]: To write this test I used test [ProtobufAccountId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/AccountIdTest.cc#L125) from `AccountIdTest`.

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  // When
+  ContractId contractId = ContractId::fromProtobuf(testProtoContractId);
+  
+  // Then
+  EXPECT_EQ(contractId.getShardNum(), testShardNum);
+  EXPECT_EQ(contractId.getRealmNum(), testRealmNum);
+  EXPECT_EQ(contractId.getContractNum(), testContractNum);
+}
+
+TEST_F(SerializationAndDeserializationTests, SerializeTransactionIdToProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:122]: To write this test I used test [ProtobufTransactionId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionIdTest.cc#L51) from [TransactionIdTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionIdTest.cc).

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  const auto protoTimestampPtr = std::unique_ptr<proto::Timestamp>(internal::TimestampConverter::toProtobuf(now));
+
+  // Then
+  EXPECT_EQ(static_cast<uint64_t>(protoTransactionIdPtr->accountid().shardnum()), testAccountId.getShardNum());
+  EXPECT_EQ(static_cast<uint64_t>(protoTransactionIdPtr->accountid().realmnum()), testAccountId.getRealmNum());
+  EXPECT_EQ(static_cast<uint64_t>(protoTransactionIdPtr->accountid().accountnum()), testAccountId.getAccountNum());
+  EXPECT_EQ(protoTransactionIdPtr->transactionvalidstart().seconds(), protoTimestampPtr->seconds());
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeTransactionIdFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:141]: To write this test I used test [ProtobufTransactionId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionIdTest.cc#L51) from [TransactionIdTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionIdTest.cc).

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+  const TransactionId transactionId = TransactionId::fromProtobuf(testProtoTransactionId);
+
+  // Then
+  EXPECT_EQ(transactionId.getAccountId().getAccountNum(), testAccountId.getAccountNum());
+  EXPECT_EQ(transactionId.getAccountId().getRealmNum(), testAccountId.getRealmNum());
+  EXPECT_EQ(transactionId.getAccountId().getShardNum(), testAccountId.getShardNum());
+  EXPECT_EQ(transactionId.getValidTransactionTime(), now);
+}
+
+TEST_F(SerializationAndDeserializationTests, SerializeTransferToProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:161]: To write this test I used test [ProtoTransferTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransferTest.cc#L33) from [TransferTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransferTest.cc).

File: sdk/tests/SerializationAndDeserializationTests.cc [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+  // When
+  const auto protoAccountAmountPtr = std::unique_ptr<proto::AccountAmount>(testTransfer.toProtobuf());
+  
+  // Then
+  EXPECT_EQ(protoAccountAmountPtr->accountid().accountnum(), testAccountId.getAccountNum());
+  EXPECT_EQ(protoAccountAmountPtr->amount(), testAmount);
+  EXPECT_FALSE(protoAccountAmountPtr->is_approval());
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeTransferFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/SerializationAndDeserializationTests.cc:182]: To write this test I used test [ProtoTransferTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransferTest.cc#L33) from [TransferTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransferTest.cc).

File: sdk/tests/DeserializationTests.cc [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+  const Hbar mBalance = Hbar(100LL);
+  const AccountId mAccountId = AccountId(0ULL, 0ULL, 10ULL);
+  const AccountId mAccountIdFrom = AccountId(4ULL);
+  const AccountId mAccountIdTo = AccountId(3ULL);
+  const std::unique_ptr<ECDSAPrivateKey> mPrivateKey = ECDSAPrivateKey::generatePrivateKey();
+  const std::shared_ptr<PublicKey> mPublicKeyFromPrivate = mPrivateKey->getPublicKey();
+  const std::shared_ptr<ECDSAPublicKey> mPublicKeyFromString = ECDSAPublicKey::fromString(mPublicKeyFromPrivate->toString());
+};
+
+TEST_F(DeserializationTests, DeserializeExchangeRateFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/DeserializationTests.cc:75]: To write this test I used test [ProtobufExchangeRate](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ExchangeRateTest.cc#L43) from [ExchangeRateTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/ExchangeRateTest.cc).

File: sdk/tests/DeserializationTests.cc [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  ExchangeRate exchangeRate = ExchangeRate::fromProtobuf(*testProtoExchangeRate);
+  
+  // Then
+  EXPECT_EQ(exchangeRate.getCurrentExchangeRate(), testCents / testHbar);
+  EXPECT_TRUE(exchangeRate.getExpirationTime().has_value());
+  EXPECT_EQ(exchangeRate.getExpirationTime().value().time_since_epoch().count(),
+            internal::TimestampConverter::fromProtobuf(*testProtoExchangeRateSecs).time_since_epoch().count());
+}
+
+TEST_F(DeserializationTests, DeserializeExchangeRateSetFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/DeserializationTests.cc:99]: To write this test I used test [ProtobufExchangeRate](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ExchangeRateTest.cc#L43) from [ExchangeRateTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/ExchangeRateTest.cc).

File: sdk/tests/DeserializationTests.cc [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+  // When
+  ExchangeRateSet exchangeRateSet = ExchangeRateSet::fromProtobuf(testProtoExchangeRateSet);
+  
+  // Then
+  EXPECT_FALSE(exchangeRateSet.getCurrentExchangeRate().has_value());
+  EXPECT_TRUE(exchangeRateSet.getNextExchangeRate().has_value());
+  EXPECT_EQ(exchangeRateSet.getNextExchangeRate().value().getCurrentExchangeRate(), testCents * testCents / testHbar * testHbar);
+}
+
+TEST_F(DeserializationTests, DeserializeTransactionReceiptFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/DeserializationTests.cc:122]: To write this test I used test [ProtobufTransactionReceipt](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionReceiptTest.cc#L46) from [TransactionReceiptTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionReceiptTest.cc).

File: sdk/tests/DeserializationTests.cc [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+  EXPECT_EQ(txRx.getExchangeRates()->getCurrentExchangeRate()->getExpirationTime(),
+            std::chrono::system_clock::time_point(std::chrono::seconds(secs)));
+  EXPECT_TRUE(txRx.getExchangeRates()->getNextExchangeRate().has_value());
+  EXPECT_EQ(txRx.getExchangeRates()->getNextExchangeRate()->getCurrentExchangeRate(), value / value);
+  EXPECT_TRUE(txRx.getExchangeRates()->getNextExchangeRate()->getExpirationTime().has_value());
+  EXPECT_EQ(txRx.getExchangeRates()->getNextExchangeRate()->getExpirationTime(),
+            std::chrono::system_clock::time_point(std::chrono::seconds(secs)));
+}
+
+TEST_F(DeserializationTests, DeserializeTransactionRecordFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/DeserializationTests.cc:161]: To write this test I used test [ProtobufTransactionRecord](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionRecordTest.cc#L34) from [TransactionRecordTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionRecordTest.cc).

File: sdk/tests/DeserializationTests.cc [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+  EXPECT_EQ(txRecord.getTransactionMemo(), testTxMemo);
+  EXPECT_EQ(txRecord.getTransactionFee(), testTxFee);
+  EXPECT_FALSE(txRecord.getTransferList().empty());
+  EXPECT_EQ(txRecord.getTransferList().at(0).getAccountId(), testAccountIdFrom);
+  EXPECT_EQ(txRecord.getTransferList().at(0).getAmount().toTinybars(), -testTransferAmount);
+  EXPECT_EQ(txRecord.getTransferList().at(1).getAccountId(), testAccountIdTo);
+  EXPECT_EQ(txRecord.getTransferList().at(1).getAmount().toTinybars(), testTransferAmount);
+}
+
+TEST_F(DeserializationTests, DeserializeTransactionResponseFromProtobufTest)
```
Comment (deyanzz) [inline sdk/tests/DeserializationTests.cc:210]: To write this test I used test [ProtobufTransactionResponse](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionResponseTest.cc#L31) from [TransactionResponseTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionResponseTest.cc).

File: sdk/tests/DeserializationTests.cc [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+  [[nodiscard]] inline const uint64_t getTestRealmNum() const { return mRealmNum; }
+  [[nodiscard]] inline const uint64_t getTestContractNum() const { return mContractNum; }
+  [[nodiscard]] inline const int32_t  getTestCents() const { return cents; }
+  [[nodiscard]] inline const int32_t  getTestHbar() const { return hbar; }
+  [[nodiscard]] inline const uint64_t getTestSeconds() const { return seconds; }
+  [[nodiscard]] inline const Hbar     getTestBalance() const { return mBalance; }
+  [[nodiscard]] inline const AccountId& getTestAccountId() const { return mAccountId; }
+  [[nodiscard]] inline const AccountId& getTestAccountIdFrom() const { return mAccountIdFrom; }
+  [[nodiscard]] inline const AccountId& getTestAccountIdTo() const { return mAccountIdTo; }
+  [[nodiscard]] inline const std::shared_ptr<ECDSAPublicKey>& getTestECDSAPublicKey() const { return mPublicKeyFromString; }
```
Comment (rwalworth) [inline sdk/tests/DeserializationTests.cc:57]: This doesn't look like it follows our formatting outlined in our `.clang-format` file. You should be able to set up your IDE to auto-format on save or on `git commit`.

File: sdk/tests/DeserializationTests.cc [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+
+private:
+  const uint64_t mShardNum = 1;
+  const uint64_t mRealmNum = 2;
+  const uint64_t mContractNum = 3;
+  const int32_t cents = 2;
+  const int32_t hbar = 1;
+  const uint64_t seconds = 100ULL;
```
Comment (rwalworth) [inline sdk/tests/DeserializationTests.cc:65]: Can these variables follow our `m<NAME>` variable naming convention?

File: sdk/tests/SerializationTests.cc [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+  // Given
+  const std::shared_ptr<ECDSAPublicKey> testECDSAPublicKey = getTestECDSAPublicKey();
+
+  // // When
+  const std::unique_ptr<proto::Key> protobufECDSAPublicKey = testECDSAPublicKey.get()->toProtobuf();
+  
+  // Then
+  EXPECT_NE(protobufECDSAPublicKey.get(), nullptr);
+  EXPECT_TRUE(protobufECDSAPublicKey.get()->has_ecdsa_secp256k1());
+}
```
Comment (rwalworth) [inline sdk/tests/SerializationTests.cc:52]: You should test serializing a `EDD25519PrivateKey` as well.

**Review by rwalworth — CHANGES_REQUESTED** [review]
It seems like a lot of the tests that you have written here are more or less copies of tests that already existed. I don't see a point in replicating those tests; any additional testing that should happen with those objects should just be put into those tests that already exist. Also, I think it may be better to have tests involving a specific object to be contained in a test file for that object as opposed to dividing the suites up by function (e.g. serialization, deserialization, etc.).

File: sdk/tests/DeserializationTests.cc (hunk 15 above)
Comment (deyanzz) [reply sdk/tests/DeserializationTests.cc]: Fixed in https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/e201bc6d2c798a8aa3b1221be2d924bd8334ecda

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 1 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Test `SerializeAccountIdToProtobufTest` & `DeserializeAccountIdFromProtobufTest` moved into `AccountIdTest.cc` in commit [11ed0182be6d48d340a9897994290084e66cf20e].(https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/11ed0182be6d48d340a9897994290084e66cf20e)

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 2 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Tests `SerializeContractIdТoProtobufTest` & `DeserializeContractIdFromProtobufTest` moved into `ContractIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/42c4b74b470cbb18745489aebed04b87c7b75235.

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 3 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Test `DeserializeContractIdFromProtobufTest` moved into `ContractIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/42c4b74b470cbb18745489aebed04b87c7b75235.

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 4 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Test `SerializeAccountIdToProtobufTest` moved into `AccountIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/11ed0182be6d48d340a9897994290084e66cf20e.

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 5 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Test `SerializeTransactionIdToProtobufTest` moved into `TransactionIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/672f803dcd2b2fc651dac7cd9f39e08cef329fff.

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 6 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Test `DeserializeTransactionIdFromProtobufTest` moved into `TransactionIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/672f803dcd2b2fc651dac7cd9f39e08cef329fff.

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 7 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Test `SerializeTransferToProtobufTest` into `TransferTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/abcec567906e347c29c9a94934ce4f14d47963d5.

File: sdk/tests/SerializationAndDeserializationTests.cc (hunk 8 above)
Comment (deyanzz) [reply sdk/tests/SerializationAndDeserializationTests.cc]: Test `DeserializeTransferToProtobufTest` into `TransferTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/abcec567906e347c29c9a94934ce4f14d47963d5.

File: sdk/tests/DeserializationTests.cc (hunk 9 above)
Comment (deyanzz) [reply sdk/tests/DeserializationTests.cc]: Test `DeserializeExchangeRateFromProtobufTest` moved into `ExchangeRateTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/ce0137a5c63aec2af768426d458144e413528ca6.

File: sdk/tests/DeserializationTests.cc (hunk 10 above)
Comment (deyanzz) [reply sdk/tests/DeserializationTests.cc]: Test `DeserializeExchangeRateSetFromProtobufTest ` moved into `ExchangeRateTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/a3879f92d8fa2ce3ce92699b7c572a84fbaca461.

File: sdk/tests/DeserializationTests.cc (hunk 11 above)
Comment (deyanzz) [reply sdk/tests/DeserializationTests.cc]: Removed test `DeserializeTransactionReceiptFromProtobufTest` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/8f2579891c1a4a154cdf7b967d8ad881cd0791cc.

File: sdk/tests/DeserializationTests.cc (hunk 13 above)
Comment (deyanzz) [reply sdk/tests/DeserializationTests.cc]: Remove test `DeserializeTransactionResponseFromProtobufTest` from `DeserializationTests.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/24e9524b0d8a82ce38fd17d6a04f8255a3d5e480.

File: sdk/tests/DeserializationTests.cc (hunk 12 above)
Comment (deyanzz) [reply sdk/tests/DeserializationTests.cc]: Remove test `DeserializeTransactionRecordFromProtobufTest ` from `DeserializationTests.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/679c14b787e9d21e506dfa2add686593a61f7863.

File: sdk/tests/SerializationTests.cc (hunk 16 above)
Comment (deyanzz) [reply sdk/tests/SerializationTests.cc]: I double-checked it, but it seems `ED25519PrivateKey` doesn't have method `toProtobuf` and can not be serialized. Could you check again and tell me if I'm wrong?

Comment (rwalworth) [reply sdk/tests/SerializationTests.cc]: Sorry, I meant `ED25519PublicKey`

**Review by rwalworth — APPROVED** [review]

**Comment by deyanzz** [comment]
> It seems like a lot of the tests that you have written here are more or less copies of tests that already existed. I don't see a point in replicating those tests; any additional testing that should happen with those objects should just be put into those tests that already exist. Also, I think it may be better to have tests involving a specific object to be contained in a test file for that object as opposed to dividing the suites up by function (e.g. serialization, deserialization, etc.).

I was my fault I didn't communicate well that approach with you in the beginning. 
I removed the 3 files for separate testing (`SerializationAndDeserializationTests.cc`, `SerializationTests.cc` and `DeserializationTests.cc`) and moved all of the tests in the corresponding `*Test.cc` files. 
I agree it's better to stick tests involving a specific object to be contained in a test file for that object, so I did it already. Thanks for the advice!

**Review by rwalworth — APPROVED** [review]
LGTM

**Pull request merged**